src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1297 @@
+                poly, (<fmpz_poly>obj).val, self.mod.val
+            )
+            return 0
+
+
+        obj = any_as_fmpz(obj)
+        if obj is NotImplemented:
+            return NotImplemented
+        fmpz_mod_poly_set_fmpz(
+            poly, (<fmpz>obj).val, self.mod.val
+        )
+        return 0
+
+    cdef any_as_fmpz_mod_poly(self, obj):
+        # Convert fmpz_mod_poly
+        if typecheck(obj, fmpz_mod_poly):
+            if self != (<fmpz_mod_poly>obj).ctx:
+                raise ValueError("moduli must match")
+            return obj
+        
+        cdef fmpz_mod_poly res
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        check = self.set_any_as_fmpz_mod_poly(res.val, obj)
+        if check is NotImplemented:
+            return NotImplemented
+        
+        res.ctx = self
+        return res
+
+    def __eq__(self, other):
+        # Most often, we expect both `fmpz_mod_poly` to be pointing 
+        # to the same ctx, so this seems the fastest way to check
+        if self is other:
+            return True
+        
+        # If they're not the same object in memory, they may have the
+        # same modulus, which is good enough
+        if typecheck(other, fmpz_mod_poly_ctx):
+            return self.mod == (<fmpz_mod_poly_ctx>other).mod
+        return False
+    
+    def __hash__(self):
+        return hash(self.modulus())
+
+    def __str__(self):
+        return f"Context for fmpz_mod_poly with modulus: {self.modulus()}"
+
+    def __repr__(self):
+        return f"fmpz_mod_poly_ctx({self.modulus()})"
+
+    def __call__(self, val):
+        return fmpz_mod_poly(val, self)
+
+
+cdef class fmpz_mod_poly(flint_poly):
+    """
+    """
+    def __cinit__(self):
+        self.initialized = False
+
+    def __dealloc__(self):
+        if self.initialized:
+            fmpz_mod_poly_clear(self.val, self.ctx.mod.val)
+
+    def __init__(self, val, ctx):
+        if not typecheck(ctx, fmpz_mod_poly_ctx):
+            raise TypeError
+        self.ctx = ctx
+        self.ctx.set_any_as_fmpz_mod_poly(self.val, val)
+        self.initialized = True
+
+    def __pos__(self):
+        return self
+
+    def __neg__(self):
+        cdef fmpz_mod_poly res
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = self.ctx
+        fmpz_mod_poly_neg(res.val, self.val, self.ctx.mod.val)
+        return res
+
+    def __add__(self, other):
+        cdef fmpz_mod_poly res
+        other = self.ctx.any_as_fmpz_mod_poly(other)
+        if other is NotImplemented:
+            return other
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        fmpz_mod_poly_add(
+            res.val, self.val, (<fmpz_mod_poly>other).val, self.ctx.mod.val
+        )
+        res.ctx = self.ctx
+        return res
+
+    def __radd__(self, other):
+        return self.__add__(other)
+
+    @staticmethod
+    def _sub_(left, right):
+        cdef fmpz_mod_poly res
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
```

Comment:
We shouldn't call `__new__` here before handling the error cases and alternative returns below.